File: csharp/TestTimeSeries/TestTimeSeries.cs
using System;
using Perst;


public class TestTimeSeries { 
    public struct Quote : TimeSeriesTick 
    { 
        public int   timestamp;
        public float low;
        public float high;
        public float open;
        public float close;
        public int   volume;

        public long Time 
        { 
            get 
            { 
                return getTicks(timestamp);
            }
        }
    }
    
    public class QuoteBlock : TimeSeriesBlock 
    {
        private Quote[] quotes;
        
        public const int N_ELEMS_PER_BLOCK = 100;

        public override TimeSeriesTick this[int i] 
        {
            get 
            {
                return quotes[i];
            }
            set 
            {
                quotes[i] = (Quote)value;
            }
        }

        public override Array Ticks 
        {
            get 
            {
                return quotes;
            }
        }

        public QuoteBlock() 
        {
            quotes = new Quote[N_ELEMS_PER_BLOCK];
        }
    }

    class Stock : Persistent { 
        public string     name;
        public TimeSeries quotes;
    }

    const int nElements = 1000000;
    const int pagePoolSize = 32*1024*1024;

    static public void Main(string[] args) {   
        Stock stock;
        int i;

        Storage db = StorageFactory.Instance.CreateStorage();
        db.Open("testts.dbs", pagePoolSize);
        FieldIndex stocks = (FieldIndex)db.Root;
        if (stocks == null) { 
            stocks = db.CreateFieldIndex(typeof(Stock), "name", true);
            stock = new Stock();
            stock.name = "BORL";
            stock.quotes = db.CreateTimeSeries(typeof(QuoteBlock), QuoteBlock.N_ELEMS_PER_BLOCK*TICKS_PER_SECOND*2);
            stocks.Put(stock);
            db.Root = stocks;
        } else { 
            stock = (Stock)stocks["BORL"];
        }
        Random rand = new Random(2004);
        DateTime start = DateTime.Now;
        int time = getSeconds(start) - nElements;
        for (i = 0; i < nElements; i++) { 
            Quote quote = new Quote();        
            quote.timestamp = time + i;
            quote.open = (float)rand.Next(10000)/100;
            quote.close = (float)rand.Next(10000)/100;
            quote.high = Math.Max(quote.open, quote.close);
            quote.low = Math.Min(quote.open, quote.close);
            quote.volume = rand.Next(1000);
            stock.quotes.Add(quote);
        }
        db.Commit();
        Console.WriteLine("Elapsed time for storing " + nElements + " quotes: " 
                          + (DateTime.Now - start));
        
        rand = new Random(2004);
        start = DateTime.Now;
        i = 0;
        foreach (Quote quote in stock.quotes) 
        {
            Assert.That(quote.timestamp == time + i);
            float open = (float)rand.Next(10000)/100;
            Assert.That(quote.open == open);
            float close = (float)rand.Next(10000)/100;
            Assert.That(quote.close == close);
            Assert.That(quote.high == Math.Max(quote.open, quote.close));
            Assert.That(quote.low == Math.Min(quote.open, quote.close));
            Assert.That(quote.volume == rand.Next(1000));
            i += 1;
        }
        Assert.That(i == nElements);
        Console.WriteLine("Elapsed time for extracting " + nElements + " quotes: " 
                           + (DateTime.Now - start));
                 
        Assert.That(stock.quotes.Count == nElements);
        
        
        long from = getTicks(time+1000);
        int count = 1000;
        start = DateTime.Now;
        i = 0;
        foreach (Quote quote in stock.quotes.Range(new DateTime(from), new DateTime(from + count*TICKS_PER_SECOND), IterationOrder.DescentOrder)) {
            Assert.That(quote.timestamp == time + 1000 + count - i);
            i += 1;
        }
        Assert.That(i == count+1);
        Console.WriteLine("Elapsed time for extracting " + i + " quotes: " + (DateTime.Now - start));

        start = DateTime.Now;
        long n = stock.quotes.Remove(stock.quotes.FirstTime, stock.quotes.LastTime);
        Assert.That(n == nElements);
        Console.WriteLine("Elapsed time for removing " + nElements + " quotes: " 
                           + (DateTime.Now - start));

        Assert.That(stock.quotes.Count == 0);
        
        db.Close();
    }


    const long TICKS_PER_SECOND = 10000000L;

    static DateTime baseDate = new DateTime(1970, 1, 1);
    
    static int getSeconds(DateTime dt) 
    {
        return (int)((dt.Ticks - baseDate.Ticks) / TICKS_PER_SECOND);
    }

    static long getTicks(int seconds) 
    {
        return baseDate.Ticks + seconds * TICKS_PER_SECOND;
    }

}

File: java/tst/TestTimeSeries.java
import org.garret.perst.*;

import java.util.*;

public class TestTimeSeries { 
    public static class Quote implements TimeSeriesTick { 
        int   timestamp;
        float low;
        float high;
        float open;
        float close;
        int   volume;

        public long getTime() { 
            return (long)timestamp*1000;
        }
    }
    
    public static class QuoteBlock extends TimeSeriesBlock {
        private Quote[] quotes;
        
        static final int N_ELEMS_PER_BLOCK = 100;

        public TimeSeriesTick[] getTicks() { 
            if (quotes == null) { 
                quotes = new Quote[N_ELEMS_PER_BLOCK];
                for (int i = 0; i < N_ELEMS_PER_BLOCK; i++) { 
                    quotes[i] = new Quote();
                }
            }
            return quotes;
        }
    }

    static class Stock extends Persistent { 
        String     name;
        TimeSeries quotes;
    }

    final static int nElements = 10000000;
    final static int pagePoolSize = 32*1024*1024;

    static public void main(String[] args) throws Exception {   
        Stock stock;
        int i;

        Storage db = StorageFactory.getInstance().createStorage();
        db.open("testts.dbs", pagePoolSize);
        FieldIndex stocks = (FieldIndex)db.getRoot();
        if (stocks == null) { 
            stocks = db.createFieldIndex(Stock.class, "name", true);
            stock = new Stock();
            stock.name = "BORL";
            stock.quotes = db.createTimeSeries(QuoteBlock.class, (long)QuoteBlock.N_ELEMS_PER_BLOCK*1000*2);
            stocks.put(stock);
            db.setRoot(stocks);
        } else { 
            stock = (Stock)stocks.get("BORL");
        }
        Random rand = new Random(2004);
        long start = System.currentTimeMillis();
        int time = (int)(start/1000) - nElements;
        for (i = 0; i < nElements; i++) { 
            Quote quote = new Quote();        
            quote.timestamp = time + i;
            quote.open = (float)rand.nextInt(10000)/100;
            quote.close = (float)rand.nextInt(10000)/100;
            quote.high = Math.max(quote.open, quote.close);
            quote.low = Math.min(quote.open, quote.close);
            quote.volume = rand.nextInt(1000);
            stock.quotes.add(quote);
        }
        db.commit();
        System.out.println("Elapsed time for storing " + nElements + " quotes: " 
                           + (System.currentTimeMillis() - start) + " milliseconds");
        
        rand.setSeed(2004);
        start = System.currentTimeMillis();
        Iterator iterator = stock.quotes.iterator();
        for (i = 0; iterator.hasNext(); i++) { 
            Quote quote = (Quote)iterator.next();
            Assert.that(quote.timestamp == time + i);
            float open = (float)rand.nextInt(10000)/100;
            Assert.that(quote.open == open);
            float close = (float)rand.nextInt(10000)/100;
            Assert.that(quote.close == close);
            Assert.that(quote.high == Math.max(quote.open, quote.close));
            Assert.that(quote.low == Math.min(quote.open, quote.close));
            Assert.that(quote.volume == rand.nextInt(1000));
        }
        Assert.that(i == nElements);
        System.out.println("Elapsed time for extracting " + nElements + " quotes: " 
                           + (System.currentTimeMillis() - start) + " milliseconds");
                 
        Assert.that(stock.quotes.size() == nElements);
        
        
        long from = (long)(time+1000)*1000;
        int count = 1000;
        start = System.currentTimeMillis();
        iterator = stock.quotes.iterator(new Date(from), new Date(from + count*1000), false);
        for (i = 0; iterator.hasNext(); i++) { 
            Quote quote = (Quote)iterator.next();
            Assert.that(quote.timestamp == time + 1000 + count - i);
        }
        Assert.that(i == count+1);
        System.out.println("Elapsed time for extracting " + i + " quotes: " 
                           + (System.currentTimeMillis() - start) + " milliseconds");

        start = System.currentTimeMillis();
        long n = stock.quotes.remove(stock.quotes.getFirstTime(), stock.quotes.getLastTime());
        Assert.that(n == nElements);
        System.out.println("Elapsed time for removing " + nElements + " quotes: " 
                           + (System.currentTimeMillis() - start) + " milliseconds");

        Assert.that(stock.quotes.size() == 0);
        
        db.close();
    }
}




